src/v/kafka/server/offset_translator.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+#include "kafka/types.h"
+#include "model/fundamental.h"
+#include "raft/configuration_manager.h"
+
+namespace kafka {
+/**
+ * Offset translator performs conversion between kafka offsets, which doesn't
+ * include raft configuration batches and log offsets which includes
+ */
+class offset_translator {
+public:
+    explicit offset_translator(const raft::configuration_manager& config_mgr)
+      : _cfg_mgr(config_mgr) {}
+
+    model::offset to_kafka_offset(model::offset o) const {
+        auto d = delta(o);
+        return model::offset(o - d);
+    }
+
+    model::offset from_kafka_offset(
+      model::offset o, model::offset hint = model::offset{}) const {
+        if (o == model::offset::max()) {
+            return o;
+        }
+        model::offset kafka_offset(o);
+        auto current = std::max(hint, kafka_offset);
+        auto next_cfg_it = _cfg_mgr.lower_bound(current);
+        // fast exit, current offset is larger than last config batch
+        // offset, just add delta
+        if (next_cfg_it == _cfg_mgr.end()) {
+            --next_cfg_it;
+            return model::offset(o) + next_cfg_it->second.idx();
+        }
+
+        auto delta = model::offset(next_cfg_it->second.idx() - 1);
+
+        while (current - delta != kafka_offset
+               || current == next_cfg_it->first) {
```

Comment:
the while condition is almost always true (current is kafka_offset so `current - delta != kafka_offset` is true for any positive delta and delta is only increasing in the while body)

let's express it explicitly:

```
if (delta == 0 && current != next_cfg_it->first) {
    return current;
}

while (true) {
    ...
```

Reply:
refactored